### ISSUE TITLE
BUG: `ELSE` merged with next statement without linebreak or space

### ISSUE THREAD

**Issue by JoernBerkefeld** [FiB3/beautyAmpCore2#3]
cloned from https://github.com/sfm-cz/beautyAmpCore/issues/4

************

It seems to happen if an `ENDIF` is directly before the `ELSE`

i've isolated the following code sample:
**problem:** note how in 2 case `ELSE InsertData` is merged together to `ELSEInsertData`

```js
const beautyAmp = require('beauty-amp-core');

/**
 *
 * @param {string} content filecontent
 * @returns {string} original string on error; formatted string on success
 */
function _beautify_beautyAmp(content) {
    // immutable at the moment:
    const ampscript = {
        capitalizeAndOrNot: true,
        capitalizeIfFor: true,
        capitalizeSet: true,
        capitalizeVar: true,
        maxParametersPerLine: 4,
    };
    // immutable at the moment:
    const editor = {
        insertSpaces: true,
        tabSize: 4,
    };
    // logs trough console only for the moment.
    const logs = {
        loggerOn: false, // <= disable logging
    };
    try {
        beautyAmp.setup(ampscript, editor, logs);
        return beautyAmp.beautify(content);
    } catch (ex) {
        console.error('File._beautify_beautyAmp:: error | ' + ex.message);
        return content;
    }
}

const content = `%%[\n\nSET @unsubMessages = 'UnsubscribeSMSMessages_DEV'\nSET @unsubEvents = 'UnsubscribeSMS_DEV'\n\nSET @num = concat('+',[MOBILE_NUMBER])\nSET @rs= RetrieveSalesforceObjects('Contact','Id,et4ae5__HasOptedOutOfMobile__c', 'MobilePhone', '=',@num) \nSET @count = rowCount(@rs)\nVAR @response\nSET @rx = '^\\+((?:9[679]\\d8[035789]\\d|6[789]\\d|5[90]\\d|42\\d|3[578]\\d|2[1-689]\\d)|9[0-58]|8[1246]|6[0-6]|5[1-8]|4[013-9]|3[0-469]|2[70]|7|1)'\nSET @prefix = concat('+',RegExMatch(@num, @rx, 1))\nIF @count>0 THEN\n\tFOR @i = 1 TO @count DO\n\t\tSET @sk = Field(Row(@rs,@i),'Id')\n\t\tIF Field(Row(@rs,1),'et4ae5__HasOptedOutOfMobile__c') == 'false' THEN\n\t\t\tSET @result = UpdateSingleSalesforceObject('Contact',@sk,'et4ae5__HasOptedOutOfMobile__c','true')\n\t\t\tIF @result == 0 THEN \n\t\t\t\tInsertData(@unsubEvents,'MobileNumber',@num,'Message',[MSG(0).NOUNS],'ContactId',@sk,'Status', 'Error Updating')\n\t\t\tELSE \n\t\t\t\tInsertData(@unsubEvents,'MobileNumber',@num,'Message',[MSG(0).NOUNS],'ContactId',@sk,'Status', 'Successfully Unsubscribed')\n\t\t\tENDIF\n\t\tELSE \n\t\t\tInsertData(@unsubEvents,'MobileNumber',@num,'Message',[MSG(0).NOUNS],'ContactId',@sk,'Status', 'Already Unsubscribed')\n\t\tENDIF\n\n\tNEXT @i\n\n\tSET @msg = Lookup(@unsubMessages,'Message','Prefix',@prefix,'Type', 'Found')\n\tIF Length(@msg) == 0 THEN\n\t\tSET @response ='You have unsubscribed and will no longer receive any messages.|'\n\tELSE \n\t\tSET @response = @msg\n\tENDIF\nELSE \n\tInsertData(@unsubEvents,'MobileNumber',@num,'Message',[MSG(0).NOUNS],'Status', 'Not Found') \n\tSET @msg = Lookup(@unsubMessages,'Message','Prefix',@prefix,'Type', 'NotFound')\n\tIF Length(@msg) == 0 THEN\n\t\tSET @response ='Sorry, we could not find you'\n\tELSE \n\t\tSET @response = @msg\n\tENDIF\nENDIF\n]%%\n%%=v(@response)=%%`;

console.log(_beautify_beautyAmp(content));
```

this will lead to this result:

```c
%%[ 
    SET @unsubMessages = 'UnsubscribeSMSMessages_DEV' 
    SET @unsubEvents = 'UnsubscribeSMS_DEV' 
    SET @num = concat('+', [MOBILE_NUMBER]) 
    SET @rs = RetrieveSalesforceObjects( 
        'Contact', 
        'Id,et4ae5__HasOptedOutOfMobile__c', 
        'MobilePhone', 
        '=', 
        @num 
    ) 
    SET @count = rowCount(@rs) 
    VAR @response 
    SET @rx = '^\+((?:9[679]\d8[035789]\d|6[789]\d|5[90]\d|42\d|3[578]\d|2[1-689]\d)|9[0-58]|8[1246]|6[0-6]|5[1-8]|4[013-9]|3[0-469]|2[70]|7|1)' 
    SET @prefix = concat('+', RegExMatch(@num, @rx, 1)) 
    IF @count>0 THEN 
        FOR @I = 1 TO @COUNT DO 
            SET @sk = Field(Row(@rs, @i), 'Id') 
            IF Field(Row(@rs,1),'et4ae5__HasOptedOutOfMobile__c') == 'false' THEN 
                SET @result = UpdateSingleSalesforceObject('Contact', @sk, 'et4ae5__HasOptedOutOfMobile__c', 'true') 
                IF @result == 0 THEN 
                    InsertData( 
                        @unsubEvents, 
                        'MobileNumber', 
                        @num, 
                        'Message', 
                        [MSG(0).NOUNS], 
                        'ContactId', 
                        @sk, 
                        'Status', 
                        'Error Updating' 
                    ) 
                ELSE 
                    InsertData( 
                        @unsubEvents, 
                        'MobileNumber', 
                        @num, 
                        'Message', 
                        [MSG(0).NOUNS], 
                        'ContactId', 
                        @sk, 
                        'Status', 
                        'Successfully Unsubscribed' 
                    ) 
                ENDIF 
            ELSEInsertData( 
                    @unsubEvents, 
                    'MobileNumber', 
                    @num, 
                    'Message', 
                    [MSG(0).NOUNS], 
                    'ContactId', 
                    @sk, 
                    'Status', 
                    'Already Unsubscribed' 
                ) 
            ENDIF 
        NEXT @i 
        SET @msg = Lookup( 
            @unsubMessages, 
            'Message', 
            'Prefix', 
            @prefix, 
            'Type', 
            'Found' 
        ) 
        IF Length(@msg) == 0 THEN 
            SET @response = 'You have unsubscribed and will no longer receive any messages.|' 
        ELSE 
            SET @response = @msg 
        ENDIF 
    ELSEInsertData( 
            @unsubEvents, 
            'MobileNumber', 
            @num, 
            'Message', 
            [MSG(0).NOUNS], 
            'Status', 
            'Not Found' 
        ) 
        SET @msg = Lookup( 
            @unsubMessages, 
            'Message', 
            'Prefix', 
            @prefix, 
            'Type', 
            'NotFound' 
        ) 
        IF Length(@msg) == 0 THEN 
            SET @response = 'Sorry, we could not find you' 
        ELSE 
            SET @response = @msg 
        ENDIF 
    ENDIF 
]%% 
%%=v(@response)=%% 
```

**Comment by FiB3** [comment]
Hello @JoernBerkefeld,
I've just published a new version 0.4.3 that should resolve not only this bug, but potentially also the comment ideas that you had in the Extension Repo. Can you try (using the package, not yet the Extension), if this resolves these bugs/features before I add it to the extension?

- [ ] This task.
- [ ] https://github.com/FiB3/beautyAmp/issues/2
- [ ] https://github.com/FiB3/beautyAmp/issues/3

**Comment by JoernBerkefeld** [comment]
alright, finally got a few hours to work on this again - sorry for the delay & hope you had a good xmas/NYE break :)

**Comment by JoernBerkefeld** [comment]
checked it with the above sample code and can confirm that it's working fine now!
will go ahead and include your new 0.4.5 in mcdev with its next release :)

**Comment by JoernBerkefeld** [comment]
hmmmm on the other hand, it now removes line breaks outside of ampscript code which isn't ideal:

```js
const content = 'line1\nline2\n%%[ SET @test = 1 ]%%';

const ampscript = {
    capitalizeAndOrNot: true,
    capitalizeIfFor: true,
    capitalizeSet: true,
    capitalizeVar: true,
    maxParametersPerLine: 4,
};
// immutable at the moment:
const editor = {
    insertSpaces: true,
    tabSize: 4,
};
// logs trough console only for the moment.
const logs = {
    loggerOn: false, // <= disable logging
};
beautyAmp.setup(ampscript, editor, logs);
const formatted = await beautyAmp.beautify(content);

await fs.writeFile('tmp/beautyAmp-result.amp', formatted);
```

**expected:**

```
line1
line2
%%[
    SET @test = 1
]%%
```

**actual:**

```
line1 line2
%%[
    SET @test = 1
]%%
``` 

Same result with a more simplified `const content = 'line1\nline2';`

and while we are talking about line breaks: if `const content = '%%[ SET @test = 1 ]%%'; is thrown at it, it returns 

```js

%%[
    SET @test = 1
]%%
``` 

--> with a leading line break that shouldn't be there.

**Comment by JoernBerkefeld** [comment]
Close to the above but still a separate issue from my POV: if `const content = '%%[ SET @test = 1 ]%%';` is thrown at it, it returns 

```js
// additional, unexpected line break here on line 1 that's not visible in GitHub because the GitHub trims it ;-)
%%[
    SET @test = 1
]%%
``` 

--> with a leading line break that shouldn't be there.

**Comment by JoernBerkefeld** [comment]
just in case the above was hard to read, this is how it looks in my failing test case now:

![image](https://github.com/FiB3/beautyAmpCore2/assets/1917227/1c8c6373-96a7-4f79-a5a5-756253e97179)

**Comment by FiB3** [comment]
Hey!
all good :) Hope you had good time too! :) 

I am happy the "Else-statement" issue is resolved.

On the next one (line1-line2 on the same line), this is caused by Prettier, which removes some of the spacing (in HTML).

The last one - (`line2\n%%[`) that's how it's working (it should have been working as such since the beginning I think). It is on purpose (to break AMPscript blocks on separate lines and improve readability).

**Comment by JoernBerkefeld** [comment]
im annoyed to say: you are right about prettier being the issue here around the line1-line2 linebreak.

the ampscript opening/closing tags: I totally get it your point but when designing HTML or text messages, line breaks do matter.
Looking at this issue further, I did notice that %%==%% behaves the way I would need it to keep formatting in place. Therefore, there is an option available to users and that's good. I would want to include an info about this different line break behavior in the docs if possible.

works as expected:

```
const content = '%%[ SET @test = 1 ]%%\nline1 %%=v(@test)=%% still line 1';
```

**Comment by JoernBerkefeld** [comment]
I suppose I should simply deactivate  `includeHtml`to fix the line1-line2 issue in SMS!


BTW: the docs don't specifically mention how to use that for beautify() - lucky your code has proper JSDocs :)

**Comment by JoernBerkefeld** [comment]
ok, that did it for me. I now selectively disable html formatting for SMS related "text"/code (transactionalSMS, mobileKeyword, mobileMessage) while keeping it on for html related stuff (types script, asset). will be released with mcdev 6.0.2 or whatever comes after 6.0.1 ;-) 

thank you for your support!
